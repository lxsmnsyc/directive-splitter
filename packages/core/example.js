import { compile } from 'dismantle';

const code = `
import { server$ } from 'my-example';

const foo = 'foo';
const bar = 'bar';

function log(prefix, suffix) {
  console.log(prefix, suffix);
}

const logPrefix = server$((prefix) => {
  if (prefix === 'foo') {
    log(prefix, foo);
  } else {
    log(prefix, bar);
  }
});
`;

const result = await compile(code, '/path/to/example.ts', {
  key: 'example',
  mode: 'server',
  env: 'development',
  directives: [
    {
      value: 'use server',
      import: {
        kind: 'named',
        name: '$$server',
        source: 'my-example',
      },
    },
  ],
  functions: [
    {
      source: {
        kind: 'named',
        name: 'server$',
        source: 'my-example',
      },
      target: {
        kind: 'named',
        name: 'server$',
        source: 'my-example/server',
      },
    },
  ],
});

console.log('ENTRY');
console.log(result.code);
console.log(result.map);

console.log('\nVIRTUALS');
for (const [name, content] of result.files) {
  console.log('FILE', name);
  console.log(content);
}

console.log('\nENTRIES');
console.log(result.entries);

console.log('\nROOTs');
console.log(result.roots);
