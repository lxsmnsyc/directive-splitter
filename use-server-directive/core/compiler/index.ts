import * as dismantle from 'dismantle';

const DEFAULT_PREFIX = '__server';
const DEFAULT_DIRECTIVE = 'use server';

export interface Options extends Pick<dismantle.Options, 'mode' | 'env'> {
  directive?: string;
  prefix?: string;
  pure?: boolean;
}

export type Output = dismantle.Output;
export type CodeOutput = dismantle.CodeOutput;

export async function compile(
  id: string,
  code: string,
  options: Options,
): Promise<Output> {
  return await dismantle.compile(id, code, {
    key: 'use-server-directive',
    mode: options.mode,
    env: options.env,
    definitions: [
      {
        type: 'block-directive',
        directive: options.directive || DEFAULT_DIRECTIVE,
        idPrefix: `/${options.prefix || DEFAULT_PREFIX}/`,
        pure: options.pure,
        target: {
          kind: 'named',
          source: `use-server-directive/${options.mode}`,
          name: '$$server',
        },
      },
    ],
  });
}
