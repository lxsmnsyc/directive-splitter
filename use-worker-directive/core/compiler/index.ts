import * as dismantle from 'dismantle';

const DEFAULT_PREFIX = '__worker';
const DEFAULT_DIRECTIVE = 'use worker';

export interface Options extends Pick<dismantle.Options, 'mode' | 'env'> {
  directive?: string;
  prefix?: string;
  pure?: boolean;
}

export type Output = dismantle.Output;
export type CodeOutput = dismantle.CodeOutput;

export async function compile(
  id: string,
  code: string,
  options: Options,
): Promise<Output> {
  return await dismantle.compile(id, code, {
    runtime: 'use-worker-directive/runtime',
    key: 'use-worker-directive',
    mode: options.mode,
    env: options.env,
    definitions: [
      {
        type: 'block-directive',
        directive: options.directive || DEFAULT_DIRECTIVE,
        idPrefix: `/${options.prefix || DEFAULT_PREFIX}/`,
        pure: options.pure,
        target: {
          kind: 'named',
          source: `use-worker-directive/${options.mode}`,
          name: '$$server',
        },
      },
    ],
  });
}
