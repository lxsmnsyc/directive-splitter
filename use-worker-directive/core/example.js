import fs from 'node:fs/promises';
import { compile } from 'use-server-directive/compiler';

console.log(
  await compile('input.ts', await fs.readFile('./input.js', 'utf-8'), {
    directive: 'use server',
    mode: 'server',
  }),
);
