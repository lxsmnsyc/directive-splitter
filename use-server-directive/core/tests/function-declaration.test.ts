import { describe, expect, it } from 'vitest';
import * as compiler from '../compiler';
import { CLIENT, ID, SERVER } from './example';

describe('FunctionDeclaration', () => {
  describe('client', () => {
    it('should transform valid server functions', async () => {
      const code = `
      async function example() {
        'use server';
        return 'foo bar';
      }
      `;
      expect(await compiler.compile(ID, code, CLIENT)).toMatchSnapshot();
    });
    it('should skip non-async server functions', async () => {
      const code = `
      function example() {
        'use server';
        return 'foo bar';
      }
      `;
      expect(await compiler.compile(ID, code, CLIENT)).toMatchSnapshot();
    });
    it('should transform valid server functions with scope', async () => {
      const code = `
      const outer = () => {
        const value = 'foo bar';
        async function example() {
          'use server';
          return value;
        }
      }
      `;
      expect(await compiler.compile(ID, code, CLIENT)).toMatchSnapshot();
    });
    it('should skip top-level values for scope', async () => {
      const code = `
      const value = 'foo bar';
      const outer = () => {
        async function example() {
          'use server';
          return value;
        }
      }
      `;
      expect(await compiler.compile(ID, code, CLIENT)).toMatchSnapshot();
    });
  });
  describe('server', () => {
    it('should transform valid server functions', async () => {
      const code = `
      async function example() {
        'use server';
        return 'foo bar';
      }
      `;
      expect(await compiler.compile(ID, code, SERVER)).toMatchSnapshot();
    });
    it('should skip non-async server functions', async () => {
      const code = `
      function example() {
        'use server';
        return 'foo bar';
      }
      `;
      expect(await compiler.compile(ID, code, SERVER)).toMatchSnapshot();
    });
    it('should transform valid server functions with scope', async () => {
      const code = `
      const outer = () => {
        const value = 'foo bar';
        async function example() {
          'use server';
          return value;
        }
      }
      `;
      expect(await compiler.compile(ID, code, SERVER)).toMatchSnapshot();
    });
    it('should skip top-level values for scope', async () => {
      const code = `
      const value = 'foo bar';
      const outer = () => {
        async function example() {
          'use server';
          return value;
        }
      }
      `;
      expect(await compiler.compile(ID, code, SERVER)).toMatchSnapshot();
    });
  });
});
