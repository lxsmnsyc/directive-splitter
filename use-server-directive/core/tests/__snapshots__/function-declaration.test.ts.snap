// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FunctionDeclaration > client > should skip non-async server functions 1`] = `
{
  "code": "function example() {
  'use server';

  return 'foo bar';
}",
  "entries": [],
  "files": Map {},
  "map": {
    "file": undefined,
    "mappings": "AACM,SAASA,OAAOA,CAAA,EAAG;EACjB,YAAY;;EACZ,OAAO,SAAS;AAClB",
    "names": [
      "example",
    ],
    "sourceRoot": undefined,
    "sources": [
      "example.ts",
    ],
    "sourcesContent": [
      "
      function example() {
        'use server';
        return 'foo bar';
      }
      ",
    ],
    "version": 3,
  },
  "roots": [],
}
`;

exports[`FunctionDeclaration > client > should skip top-level values for scope 1`] = `
{
  "code": "import { $$func as $$func_1 } from "dismantle/runtime";
import { value } from "./example.ts?use-server-directive=0.ts";
const outer = () => {
  async function example() {
    const [type_1, result_1] = await $$func_1((await import("./example.ts?use-server-directive=1.ts")).default, null)();
    if (type_1 === 2) {
      return result_1;
    }
  }
};",
  "entries": [
    "example.ts?use-server-directive=1.ts",
  ],
  "files": Map {
    "example.ts?use-server-directive=0.ts" => {
      "code": "export const value = 'foo bar';",
      "map": {
        "file": undefined,
        "mappings": "aACYA,KAAK,GAAG,SAAS",
        "names": [
          "value",
        ],
        "sourceRoot": undefined,
        "sources": [
          "example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "example.ts?use-server-directive=1.ts" => {
      "code": "import { $$server as entry_1 } from "use-server-directive/client";
export default entry_1("/__server/bb5fd6ee-0-example");",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";;AAEM,MAAMA,KAAK,GAAGA,CAAA,KAAM;EAClB,eAAeC,OAAOA,CAAA;IAAA,OAAAC,MAAA,EAAAC,QAAA,UAAAC,QAAA,0DAAAC,OAAA;IAAA,IAAAH,MAAA;MAAA,OAAAC,QAAA;IAAA;EAAA;AAIxB,CAAC",
    "names": [
      "outer",
      "example",
      "type_1",
      "result_1",
      "$$func_1",
      "default",
    ],
    "sourceRoot": undefined,
    "sources": [
      "example.ts",
    ],
    "sourcesContent": [
      "
      const value = 'foo bar';
      const outer = () => {
        async function example() {
          'use server';
          return value;
        }
      }
      ",
    ],
    "version": 3,
  },
  "roots": [],
}
`;

exports[`FunctionDeclaration > client > should transform valid server functions 1`] = `
{
  "code": "import { $$func as $$func_1 } from "dismantle/runtime";
async function example() {
  const [type_1, result_1] = await $$func_1((await import("./example.ts?use-server-directive=0.ts")).default, null)();
  if (type_1 === 2) {
    return result_1;
  }
}",
  "entries": [
    "example.ts?use-server-directive=0.ts",
  ],
  "files": Map {
    "example.ts?use-server-directive=0.ts" => {
      "code": "import { $$server as entry_1 } from "use-server-directive/client";
export default entry_1("/__server/bb5fd6ee-0-example");",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";AACM,eAAeA,OAAOA,CAAA;EAAA,OAAAC,MAAA,EAAAC,QAAA,UAAAC,QAAA,0DAAAC,OAAA;EAAA,IAAAH,MAAA;IAAA,OAAAC,QAAA;EAAA;AAAA",
    "names": [
      "example",
      "type_1",
      "result_1",
      "$$func_1",
      "default",
    ],
    "sourceRoot": undefined,
    "sources": [
      "example.ts",
    ],
    "sourcesContent": [
      "
      async function example() {
        'use server';
        return 'foo bar';
      }
      ",
    ],
    "version": 3,
  },
  "roots": [],
}
`;

exports[`FunctionDeclaration > client > should transform valid server functions with scope 1`] = `
{
  "code": "import { $$func as $$func_1 } from "dismantle/runtime";
const outer = () => {
  const value = 'foo bar';
  async function example() {
    const [type_1, result_1] = await $$func_1((await import("./example.ts?use-server-directive=0.ts")).default, null)(value);
    if (type_1 === 2) {
      return result_1;
    }
  }
};",
  "entries": [
    "example.ts?use-server-directive=0.ts",
  ],
  "files": Map {
    "example.ts?use-server-directive=0.ts" => {
      "code": "import { $$server as entry_1 } from "use-server-directive/client";
export default entry_1("/__server/bb5fd6ee-0-example");",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";AACM,MAAMA,KAAK,GAAGA,CAAA,KAAM;EAClB,MAAMC,KAAK,GAAG,SAAS;EACvB,eAAeC,OAAOA,CAAA;IAAA,OAAAC,MAAA,EAAAC,QAAA,UAAAC,QAAA,0DAAAC,OAAA,QADhBL,KAAK;IAAA,IAAAE,MAAA;MAAA,OAAAC,QAAA;IAAA;EAAA;AAKb,CAAC",
    "names": [
      "outer",
      "value",
      "example",
      "type_1",
      "result_1",
      "$$func_1",
      "default",
    ],
    "sourceRoot": undefined,
    "sources": [
      "example.ts",
    ],
    "sourcesContent": [
      "
      const outer = () => {
        const value = 'foo bar';
        async function example() {
          'use server';
          return value;
        }
      }
      ",
    ],
    "version": 3,
  },
  "roots": [],
}
`;

exports[`FunctionDeclaration > server > should skip non-async server functions 1`] = `
{
  "code": "function example() {
  'use server';

  return 'foo bar';
}",
  "entries": [],
  "files": Map {},
  "map": {
    "file": undefined,
    "mappings": "AACM,SAASA,OAAOA,CAAA,EAAG;EACjB,YAAY;;EACZ,OAAO,SAAS;AAClB",
    "names": [
      "example",
    ],
    "sourceRoot": undefined,
    "sources": [
      "example.ts",
    ],
    "sourcesContent": [
      "
      function example() {
        'use server';
        return 'foo bar';
      }
      ",
    ],
    "version": 3,
  },
  "roots": [],
}
`;

exports[`FunctionDeclaration > server > should skip top-level values for scope 1`] = `
{
  "code": "import { $$func as $$func_1 } from "dismantle/runtime";
import { value } from "./example.ts?use-server-directive=0.ts";
const outer = () => {
  async function example() {
    const [type_1, result_1] = await $$func_1((await import("./example.ts?use-server-directive=2.ts")).default, null)();
    if (type_1 === 2) {
      return result_1;
    }
  }
};",
  "entries": [
    "example.ts?use-server-directive=2.ts",
  ],
  "files": Map {
    "example.ts?use-server-directive=0.ts" => {
      "code": "export const value = 'foo bar';",
      "map": {
        "file": undefined,
        "mappings": "aACYA,KAAK,GAAG,SAAS",
        "names": [
          "value",
        ],
        "sourceRoot": undefined,
        "sources": [
          "example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "example.ts?use-server-directive=1.ts" => {
      "code": "import { value } from "./example.ts?use-server-directive=0.ts";
export default (async function () {
  try {
    return [2, value, null];
  } catch (error_1) {
    return [4, error_1];
  }
  return [3];
});",
      "map": {
        "file": undefined,
        "mappings": ";;;IAKU,WAAOA,KAAK;EAAC,SAAAC,OAAA;IAAA,WAAAA,OAAA;EAAA;EAAA;AAAA",
        "names": [
          "value",
          "error_1",
        ],
        "sourceRoot": undefined,
        "sources": [
          "example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "example.ts?use-server-directive=2.ts" => {
      "code": "import { $$server as entry_1 } from "use-server-directive/server";
import root_1 from "./example.ts?use-server-directive=1.ts";
export default entry_1("/__server/bb5fd6ee-0-example", root_1);",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";;AAEM,MAAMA,KAAK,GAAGA,CAAA,KAAM;EAClB,eAAeC,OAAOA,CAAA;IAAA,OAAAC,MAAA,EAAAC,QAAA,UAAAC,QAAA,0DAAAC,OAAA;IAAA,IAAAH,MAAA;MAAA,OAAAC,QAAA;IAAA;EAAA;AAIxB,CAAC",
    "names": [
      "outer",
      "example",
      "type_1",
      "result_1",
      "$$func_1",
      "default",
    ],
    "sourceRoot": undefined,
    "sources": [
      "example.ts",
    ],
    "sourcesContent": [
      "
      const value = 'foo bar';
      const outer = () => {
        async function example() {
          'use server';
          return value;
        }
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "example.ts?use-server-directive=1.ts",
  ],
}
`;

exports[`FunctionDeclaration > server > should transform valid server functions 1`] = `
{
  "code": "import { $$func as $$func_1 } from "dismantle/runtime";
async function example() {
  const [type_1, result_1] = await $$func_1((await import("./example.ts?use-server-directive=1.ts")).default, null)();
  if (type_1 === 2) {
    return result_1;
  }
}",
  "entries": [
    "example.ts?use-server-directive=1.ts",
  ],
  "files": Map {
    "example.ts?use-server-directive=0.ts" => {
      "code": "export default (async function () {
  try {
    return [2, 'foo bar', null];
  } catch (error_1) {
    return [4, error_1];
  }
  return [3];
});",
      "map": {
        "file": undefined,
        "mappings": ";;IAGQ,WAAO,SAAS;EAAC,SAAAA,OAAA;IAAA,WAAAA,OAAA;EAAA;EAAA;AAAA",
        "names": [
          "error_1",
        ],
        "sourceRoot": undefined,
        "sources": [
          "example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "example.ts?use-server-directive=1.ts" => {
      "code": "import { $$server as entry_1 } from "use-server-directive/server";
import root_1 from "./example.ts?use-server-directive=0.ts";
export default entry_1("/__server/bb5fd6ee-0-example", root_1);",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";AACM,eAAeA,OAAOA,CAAA;EAAA,OAAAC,MAAA,EAAAC,QAAA,UAAAC,QAAA,0DAAAC,OAAA;EAAA,IAAAH,MAAA;IAAA,OAAAC,QAAA;EAAA;AAAA",
    "names": [
      "example",
      "type_1",
      "result_1",
      "$$func_1",
      "default",
    ],
    "sourceRoot": undefined,
    "sources": [
      "example.ts",
    ],
    "sourcesContent": [
      "
      async function example() {
        'use server';
        return 'foo bar';
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "example.ts?use-server-directive=0.ts",
  ],
}
`;

exports[`FunctionDeclaration > server > should transform valid server functions with scope 1`] = `
{
  "code": "import { $$func as $$func_1 } from "dismantle/runtime";
const outer = () => {
  const value = 'foo bar';
  async function example() {
    const [type_1, result_1] = await $$func_1((await import("./example.ts?use-server-directive=1.ts")).default, null)(value);
    if (type_1 === 2) {
      return result_1;
    }
  }
};",
  "entries": [
    "example.ts?use-server-directive=1.ts",
  ],
  "files": Map {
    "example.ts?use-server-directive=0.ts" => {
      "code": "export default (async function (value) {
  try {
    return [2, value, null];
  } catch (error_1) {
    return [4, error_1];
  }
  return [3];
});",
      "map": {
        "file": undefined,
        "mappings": "gCAEcA,KAAK;EAAA;IAGT,WAAOA,KAAK;EAAC,SAAAC,OAAA;IAAA,WAAAA,OAAA;EAAA;EAAA;AAAA",
        "names": [
          "value",
          "error_1",
        ],
        "sourceRoot": undefined,
        "sources": [
          "example.ts",
        ],
        "sourcesContent": [
          null,
        ],
        "version": 3,
      },
    },
    "example.ts?use-server-directive=1.ts" => {
      "code": "import { $$server as entry_1 } from "use-server-directive/server";
import root_1 from "./example.ts?use-server-directive=0.ts";
export default entry_1("/__server/bb5fd6ee-0-example", root_1);",
      "map": {
        "file": undefined,
        "mappings": "",
        "names": [],
        "sourceRoot": undefined,
        "sources": [],
        "sourcesContent": [],
        "version": 3,
      },
    },
  },
  "map": {
    "file": undefined,
    "mappings": ";AACM,MAAMA,KAAK,GAAGA,CAAA,KAAM;EAClB,MAAMC,KAAK,GAAG,SAAS;EACvB,eAAeC,OAAOA,CAAA;IAAA,OAAAC,MAAA,EAAAC,QAAA,UAAAC,QAAA,0DAAAC,OAAA,QADhBL,KAAK;IAAA,IAAAE,MAAA;MAAA,OAAAC,QAAA;IAAA;EAAA;AAKb,CAAC",
    "names": [
      "outer",
      "value",
      "example",
      "type_1",
      "result_1",
      "$$func_1",
      "default",
    ],
    "sourceRoot": undefined,
    "sources": [
      "example.ts",
    ],
    "sourcesContent": [
      "
      const outer = () => {
        const value = 'foo bar';
        async function example() {
          'use server';
          return value;
        }
      }
      ",
    ],
    "version": 3,
  },
  "roots": [
    "example.ts?use-server-directive=0.ts",
  ],
}
`;
